### ISSUE TITLE
Use Imager-X to make custom URL requests to Imgix?

### ISSUE THREAD

**Issue by cmalven** [spacecatninja/craft-imager-x#33]
### I'm submitting a... 
- [ ] bug report
- [ ] feature request
- [x] general question

### Description
Imgix offers some URL APIs that go beyond what Imager-X is capable of doing. I'm specifically looking at the `palette` API, which offers more flexibility in the variety of colors returned than the similar `dominantColor` and `getColorPalette` options Imager provides. They're also probably a lot less demanding on the server since all of the calculation happens on Imgix's super speedy servers.

Imager already knows how to construct an Imgix URL – including adding the secure token – so it would be nice if I could at least construct the Imgix API URL for me, if not actually make the API call itself. The resulting URL would be something like `https://static.imgix.net/treefrog.jpg?w=640&h=320&fit=crop&palette=json`

Is something that could help with this available and just undocumented? I understand that it would apply almost exclusively to Imgix as typically there wouldn't be an external service to call if you aren't using Imgix, but it would be great if Imager could somehow help me out here.

Thanks!

**Comment by aelvan** [comment]
Hi,

No, currently there's nothing to help you with this unfortunately. I've actually been meaning to implement something similar to get [meta information about an image](https://docs.imgix.com/apis/url/format/fm#json), but haven't gotten around to it yet and you're the first to request it. :) 

It's not a big deal to implement, but I'm a bit torn on how/where to implement it. With the meta information part, it makes sense to have it on the transformed image model, so that you'd go:

```
{% set transformedImage = craft.imager.transformImage(asset, { width: 640, height: 320 }) %}
{% set metaInformation = transformedImage.getMetaInformation() %}
{{ metaInformation.DPIWidth }}
```

Would it make sense to do the same thing with the palette information? Or would it be preferable to have a separate template variable to get the information, like `craft.imager.imgix.getPalette(asset)`? Also, I see that it's possible to return the palette as both json and css. Is both necessary, I'm feeling like json would be the most flexible option and css a bit redundant?

**Comment by cmalven** [comment]
Thanks for the thoughtful response, @aelvan!

My concern would be that with the `transformedImage.getMetaInformation()` it isn't clear where the information is coming from, and in some cases I imagine it would be generated internally by Imager, and in other cases it would be returned from Imgix. And the information from the two might not be very similar, such as in the case of palette information.

I like that the `craft.imager.imgix.foo()` approach makes it very clear that the info is being requested form Imgix and is likely to match what Imgix describes in their docs, but I can understand why having a unified `.getMetaInformation()` might be preferable to you and I'd be okay with that if that makes more sense.

While I _personally_ haven't yet used the Imgix `css` option for color pallette, I can certainly see a use-case for it. It does address a useful situation that can't be entirely recreated through the `json` pallete, which is to create combinations of foreground/background colors that have appropriate contrast and are both pulled from the image palette. I could see that being really nice if I had a situation where I wanted to have readable text on a background and for both the text and background to match the image with very little effort. If it isn't a big deal to you either way it'd be nice to implement both, but my biggest use-case is the `json` response.

**Comment by aelvan** [comment]
Maybe I wasn't clear, I didn't mean to conceal the palette information inside that suggested `getMetaInformation` method, that was just an example on how to access it.

Anyways, I gave this some thought and went with my first inclination of adding it on the transformed image model (now available in v3.1.0). To be clear, this has only been added to the `ImgixTransformedImageModel`, so it's specific to transformed images when using the Imgix transformer. The main reason I did this, is that if I were to have a separate template variable for it, there would have to be a lot of duplicate functionality with the `transformImage` method, since you probably still want to be able to get the palette information based on the transformed image, not the original one. All that information is already available if you've done an image transform, so I felt like it would be a waste of time to have to replicate it elsewhere.

Anyways, I've updated the documentation for it [here](https://imager-x.spacecat.ninja/models.html#imgixtransformedimagemodel) and [here](https://imager-x.spacecat.ninja/usage/colors.html#when-in-imgix). As per the example, you can use it like: 

```
{% set transformedImage = craft.imager.transformImage(image, { width: 300, height: 400 }) %}
{% set palette = transformedImage.getPalette('json', 8) %}

<div>
    {% for paletteColor in palette.colors %}
        <span style="display: inline-block; width: 30px; height: 30px; background: {{ paletteColor.hex }};"></span>
    {% endfor %}
</div>
```

If you do `transformedImage.getPalette('css')` you'll get a string back with the css classes.

Hope this helps, let me know if you've any comments or questions.

**Comment by cmalven** [comment]
@aelvan This is great. Thank you, and impressive turnaround time!
